# Collaborative TMLE: which baseline variables belong in the assignment model?

This notebook estimates one average treatment effect with a TMLE whose assignment model the study
design fixes. It then runs collaborative TMLE (C-TMLE) as a data-adaptive check. Each step shows
its code, its output, and what the output tells you.
[Collaborative TMLE](../technical-reference/collaborative-tmle.md) gives the candidate paths, the
selection loss, and the fold structure.

## The applied question

Clinical and operations staff approve three baseline variables for the navigation analysis. They
confirm that each variable is measured before assignment and that none is a collider.

The operations review also certifies the queue mechanism. An encounter-ID hash sets the queue draw
before assignment, and staff cannot override it. The draw therefore changes the offer and no other
service, so the design makes it an instrument. The analysis plan leaves it out of the assignment
model before any model runs. The disjunctive cause criterion of
[VanderWeele (2019)](https://doi.org/10.1007/s10654-019-00494-6) gives this rule: exclude a
variable known to be an instrument.

The clinical review cannot certify whether readiness and social support confound the offer and the
score. The reported fit adjusts for both. C-TMLE searches all three approved variables without the
design knowledge, as a check. An assignment model that uses all three variables predicts
assignment well, and Step 8 prints its AUC. Predictive accuracy is the wrong criterion for that
model, and Step 8 shows why.

## What you will learn

| after this notebook you can | the step that shows it |
| --- | --- |
| write the protocol for a question with an approved baseline set | Step 4 |
| check that a method is available before you fit it | Step 5 |
| fit C-TMLE with explicit learners, and read its selection path | Steps 6 and 7 |
| report a TMLE whose adjustment set the design fixes | Step 8 |
| see what an instrument does to an assignment model | Step 8 |
| use C-TMLE as a check, and not as the source of the reported set | Step 8 |
| tell a selector that discriminates from one that selects nothing | Step 9 |
| read the assessment of a selected working model | Step 10 |
| run the sensitivity analysis of each fit | Step 11 |

## Why this method

| your situation | what this method buys | what it costs |
| --- | --- | --- |
| an approved baseline set whose roles the design does not settle | an assignment model selected by cross-validated loss on the targeted outcome regression, so an instrument can be left out | with `strategy="greedy"`, one assignment-model fit for each remaining variable at each stage, in each selection fold |
| the outcome regression is already good | the empty assignment model is a legitimate candidate | selecting it is not evidence that the search discriminates |

The selector chooses a nuisance model inside the approved set. It does not discover a causal
adjustment set. When the design names an instrument, leave it out of the adjustment set before any
fit. This page does that for the queue draw, and it runs the selector on all three approved
variables as a check. The table below defines the terms this notebook uses most.

| term | plain meaning | canonical definition |
| --- | --- | --- |
| estimand | the number the question asks for, written before any model is chosen | [estimands](../user-guide/estimands.md) |
| nuisance | a model the estimate needs but the question does not ask about. Here, the outcome regression Q and the assignment model g | [point-treatment TMLE](../technical-reference/point-treatment-tmle.md) |
| targeting | a small update to Q, weighted by g, that removes first-order bias | [targeting and bounds](../user-guide/methods-learners.md#targeting-and-bounds) |
| influence curve | how much each row moves the estimate. Its variance gives the standard error | [inference](../technical-reference/inference.md) |
| positivity | every kind of patient has some chance of each arm | [diagnostics](../user-guide/results-assessment.md#diagnostics) |
| instrument | a variable that changes assignment and has no other path to the outcome | [collaborative TMLE](../technical-reference/collaborative-tmle.md#what-this-solves) |

## Step 1: set up

The setup imports the learners and prints the installed `cleverly` version. Every fit below passes
its learners, fold counts, and random seed explicitly, so a rerun reproduces the stored outputs.


In [1]:
import pandas as pd
from sklearn.dummy import DummyRegressor
from sklearn.linear_model import LinearRegression, LogisticRegression

import cleverly

pd.set_option("display.width", 130)
pd.set_option("display.max_columns", 20)
print(f"cleverly {cleverly.__version__}")

cleverly 0.1.2


**What this output tells you.** The stored outputs in this notebook came from the version named
above. A different version can print different numbers.


## Step 2: the data

The data come from a synthetic law with a known answer. The generator is `make_instrument`. The code
renames its columns to the program's names, prints the first rows, and prints the true values of the
law.


In [2]:
from cleverly.datasets import make_instrument

frame, truth = make_instrument(n=2_000, seed=44)
frame = frame.rename(
    columns={
        "Y": "transition_score",
        "A": "transition_navigation",
        "W1": "baseline_readiness",
        "W2": "queue_lottery_draw",
        "W3": "social_support",
    }
)
lines = [
    f"rows and columns: {frame.shape}",
    frame.head().round(3).to_string(),
    "",
    "known values of the synthetic law:",
    *(f"  {key}: {truth[key]:.3f}" for key in ("ate", "att", "atc")),
]
print("\n".join(lines))

rows and columns: (2000, 5)
   transition_score  transition_navigation  baseline_readiness  queue_lottery_draw  social_support
0             3.472                    0.0               1.446               0.102           0.327
1             2.981                    1.0               1.136               0.824           0.579
2             2.604                    1.0              -0.379              -1.019           0.376
3             1.199                    1.0              -0.512               0.827           0.211
4             1.866                    1.0              -0.745               0.268           0.709

known values of the synthetic law:
  ate: 1.000
  att: 1.000
  atc: 1.000


**What this output tells you.** Each row is one discharge. `transition_navigation` is 1 for an
offer and 0 for usual support. The three baseline covariates are independent standard-normal
draws. The score is in synthetic units.

The score of this page has Gaussian noise, so it has no documented range. The rest of the program
records a share of the maximum score, which lies between 0 and 1. Steps 4 and 6 return to that
difference.

The effect is constant in this law, so the population `ate`, `att`, and `atc` all equal 1.000. The
three columns have separate roles in the law.

| column | role in the law | what it is in the program |
| --- | --- | --- |
| `baseline_readiness` | confounder | navigators prioritize patients ready to engage, so higher readiness raises the chance of an offer and the transition score |
| `queue_lottery_draw` | instrument | an encounter-ID hash sets a queue draw. A higher draw strongly raises the chance of an offer and has no other path to the score |
| `social_support` | outcome predictor | it moves the transition score and does not move assignment |

The queue draw is an instrument only under three conditions. The program fixes the hash before
assignment, prevents staff overrides, and verifies that the draw changes no other service. The
protocol of Step 4 records this design basis. The data cannot establish this exclusion restriction.
They cannot establish any other identification assumption of this page either.

A real program has no `truth`. Every comparison against it below is a teaching device.

## Step 3: association first

A confounder changes both who receives the offer and the outcome. The code compares the two arms
before any adjustment. It prints the mean score and the mean of each baseline covariate by arm.


In [3]:
covariates = ["baseline_readiness", "queue_lottery_draw", "social_support"]
by_arm = frame.groupby("transition_navigation")[["transition_score", *covariates]].mean()
print(by_arm.round(3))
print()
print(f"share offered navigation: {frame['transition_navigation'].mean():.3f}")
unadjusted = by_arm.loc[1.0, "transition_score"] - by_arm.loc[0.0, "transition_score"]
print(f"unadjusted difference in mean score: {unadjusted:.3f}")
print(f"population ATE:                      {truth['ate']:.3f}")

                       transition_score  baseline_readiness  queue_lottery_draw  social_support
transition_navigation                                                                          
0.0                               0.517              -0.288              -0.514          -0.035
1.0                               2.494               0.321               0.476           0.042

share offered navigation: 0.512
unadjusted difference in mean score: 1.977
population ATE:                      1.000


**What this output tells you.** The offered patients score 1.977 points higher on average, and the
true effect is 1.000. The arms differ before the offer. The mean `baseline_readiness` is 0.321
among offered patients and -0.288 among the others.

The mean `queue_lottery_draw` also differs between the arms, 0.476 against -0.514. That gap does
not bias the comparison, because the draw has no path to the score in this law. The readiness gap
does bias it. Both gaps predict the offer. The data on assignment alone cannot show which gap also
reaches the score.

## Step 4: write the protocol

A `StudyProtocol` records the scientific design before any model runs. The record gets a
fingerprint, and every result fitted from it carries that fingerprint. This page starts from
`navigation_protocol()`, the protocol of the
[shared study design](index.md#the-shared-study-design). `dataclasses.replace` adds the approved
baseline review.

In [4]:
from dataclasses import replace

from cleverly.datasets import navigation_protocol

program = navigation_protocol()
protocol = replace(
    program,
    outcome="Patient-reported transition score in synthetic units, with no documented range",
    time_zero=(
        "Discharge-home order, after baseline measurement and the queue draw, "
        "and before the navigation offer"
    ),
    intercurrent_event_handling=(
        *program.intercurrent_event_handling[:2],
        "The score has no worst value, so the protocol handles death before day 30 "
        "by the hypothetical strategy: the score had the patient not died",
    ),
    assumption_rationale=(
        "Clinical and operations review approved three baseline variables, each measured "
        "before assignment and none a collider",
        "The approved baseline variables cover the measured common causes of the offer "
        "and the score",
        "An encounter-ID hash fixes the queue draw before assignment, without staff overrides",
        *program.assumption_rationale[1:],
    ),
)
print("\n".join(protocol.summary_lines()))

causal study protocol: schema 1; b81fd2923b8e38c8
target population: Adults with a discharge-home order at a participating hospital during the enrollment period
eligibility: ['Age 18 years or older', 'Discharge home ordered at a participating hospital']
time zero: Discharge-home order, after baseline measurement and the queue draw, and before the navigation offer
treatment strategies: ['Offer standard transition navigation', 'Provide usual discharge support']
treatment versions: ['Bedside transition plan and two scheduled navigator contacts within 30 days', 'No access to the transition-navigation offer']
outcome: Patient-reported transition score in synthetic units, with no documented range
horizon: 30 days after discharge
intercurrent-event handling: ['Use the transition score regardless of readmission', 'Analyze the offer regardless of completed contacts', 'The score has no worst value, so the protocol handles death before day 30 by the hypothetical strategy: the score had the patien

**What this output tells you.** The first line gives the schema version and the fingerprint
`b81fd2923b8e38c8`. The other lines repeat each field. This page changes four fields of
`navigation_protocol()`, and [point-treatment TMLE](point-treatment-tmle.ipynb) reads the others.

| protocol field | what this page changes |
| --- | --- |
| outcome | the score is in synthetic units and has no documented range. The program records a share of the maximum score |
| intercurrent-event handling | the program scores death as the worst score, and this score has no worst value. This page uses the hypothetical strategy for death |
| time zero | the queue draw happens before the offer, so the draw is a baseline variable |
| assumption rationale | the approved baseline review, and the hash that fixes the queue draw before assignment |

Under the hypothetical strategy, a real analysis treats death as censoring. It then needs
exchangeability and positivity for death given the recorded history. This synthetic law draws no
deaths, so the choice does not change the estimate here.

Every fit on this page runs in sample. Its learners are linear models with few coefficients, so
they satisfy the Donsker condition that [CV-TMLE](../technical-reference/cv-tmle.md#what-this-solves)
names. A cross-fitted fit of a continuous outcome would also declare the outcome's range as
`q_bounds`. This score has no documented range to declare.

Two design choices have no `StudyProtocol` field. The adjustment sets are the columns of the two
designs in Step 5. The selection rule belongs to the method in Step 6. The exclusion restriction
for the queue draw has no field either. The assumption rationale records its design basis, and the
data cannot verify it.

## Step 5: design and identification

The code builds two designs on the same data. The first holds the design-based set,
`baseline_readiness` and `social_support`, as the analysis plan requires. Its fit in Step 8 is the
reported analysis. The second holds all three approved variables. C-TMLE searches that set in
Step 6, and Step 8 fits a plain TMLE on it to show the failure mode.

In this synthetic law, `baseline_readiness` alone closes the common-cause path. C-TMLE selects
terms for the assignment nuisance, and it does not revise that identification decision.

The code identifies the ATE on both designs and prints the summary of the reported one. It then
prints both adjustment sets and lists the methods that can fit the ATE.

In [5]:
from cleverly import ATE, CausalStudy, PointTreatment

design_set = ("baseline_readiness", "social_support")
approved_set = ("baseline_readiness", "queue_lottery_draw", "social_support")


def identify(adjustment):
    study = CausalStudy(
        frame,
        design=PointTreatment(
            outcome="transition_score",
            treatment="transition_navigation",
            adjustment=adjustment,
        ),
        protocol=protocol,
    )
    return study.identify(ATE(reference=0))


reported_effect = identify(design_set)
effect = identify(approved_set)
print(reported_effect.summary(protocol="fingerprint"))
print()
print(f"design-based adjustment set: {design_set}")
print(f"approved adjustment set:     {approved_set}")
available = [method.name for method in effect.available_methods() if method.available]
print(f"methods available for the ATE: {available}")

average treatment effect, E[Y^a] - E[Y^reference]
identified by explicit-adjustment: E_W[E(transition_score | transition_navigation=a, W)] - E_W[E(transition_score | transition_navigation=0, W)] for a in [1]
adjustment/history: ['baseline_readiness', 'social_support']
required nuisances: ['outcome_regression', 'treatment_mechanism']
assumptions:
  - consistency: Y = Y^a when A = a
  - no interference: one unit's potential outcome does not depend on other units' treatment assignments
  - no unmeasured confounding: Y^a is independent of A given W
  - positivity: P(transition_navigation = a | W) > 0 almost surely for every supported treatment level a in [0, 1]
causal study protocol: schema 1; b81fd2923b8e38c8

design-based adjustment set: ('baseline_readiness', 'social_support')
approved adjustment set:     ('baseline_readiness', 'queue_lottery_draw', 'social_support')
methods available for the ATE: ['tmle', 'collaborative_tmle', 'drtmle']


**What this output tells you.** The first lines show the ATE and its observed-data formula. The
formula averages the outcome regression over all patients, once with the offer and once without.
The required nuisances are the outcome regression Q and the treatment mechanism g. The summary then
lists four assumptions and prints the fingerprint of the stored protocol.

The four assumptions are those of
[point-treatment TMLE](point-treatment-tmle.ipynb#step-5-design-and-identification), which reads
each one for the program. Here no unmeasured confounding means that `baseline_readiness` and
`social_support` block every common cause of the offer and the score. The data cannot check it.

The next two lines print the two adjustment sets. The last line lists `collaborative_tmle` beside `tmle` and `drtmle`, so C-TMLE can fit this ATE.
C-TMLE in `cleverly` covers the point-treatment `ate`, `ey`, `ey1`, `ey0`, `rr`, and `or` targets.
The [technical entry](../technical-reference/collaborative-tmle.md#what-this-solves) states that
scope.

## Step 6: estimate the ATE with C-TMLE

C-TMLE runs on the approved set of all three variables. It has no knowledge that the queue draw is
an instrument, so its selection checks the design from the data. The configuration is written out
in full. Linear learners are enough here, because the outcome mean of this law is linear in the
offer and the covariates.

| setting | value | what it does |
| --- | --- | --- |
| `outcome_learner` | `LinearRegression(n_jobs=1)` | fits Q, the expected score given the offer and the covariates |
| `treatment_learner` | `LogisticRegression(max_iter=1000, random_state=44)` | fits each candidate g on the covariates that candidate uses |
| `CrossFitting(enabled=False)` | no outer folds | fits each nuisance on every row |
| `strategy="greedy"` | greedy path | at each stage, adds to g the covariate whose targeted outcome regression has the smallest penalized loss |
| `selection_folds=3` | three selection folds | cross-validates the loss that chooses the stopping point on the path |
| `selection_inner_folds=2` | two inner folds | limits the extra fits inside each selection fold |
| `Runtime(random_state=44, n_jobs=1)` | fixed seed, one process | makes the fit reproducible |

The fit runs in sample, because its learners are linear models with few coefficients. Step 4 gives
the condition that they satisfy. The synthetic score also has no documented range to declare as
`q_bounds`. A real questionnaire has a published range. With flexible learners, declare that range
as `q_bounds` on the questionnaire's own scale, and cross-fit.

With the linear learners of this page, the fold layer changes little. A review probe,
`reviews/notebook-review/probes/collaborative-tmle-final/crossfit_check.py`, refits this draw with
five outer folds and a convention range of -20 to 20. Both estimates moved by less than 0.003. The
[cross-fitting tutorial](cross-fitting.ipynb) shows the fold layer on a score whose range the
program records.

The selection folds are a separate layer, and they stay. The
[technical entry](../technical-reference/collaborative-tmle.md#the-algorithm-as-implemented)
describes both layers. The two plain TMLE fits in Step 8 use the same learners and the same
in-sample nuisances. Only the assignment model differs. The code keeps the method as
`collaborative_method`, so Step 9 can reuse it.

In [6]:
from cleverly import (
    CollaborativeTMLEMethod,
    CrossFitting,
    ModelSpec,
    Runtime,
    TMLEMethod,
)

models = ModelSpec(
    outcome_learner=LinearRegression(n_jobs=1),
    treatment_learner=LogisticRegression(max_iter=1000, random_state=44),
)
folds = CrossFitting(enabled=False)
runtime = Runtime(random_state=44, n_jobs=1)

collaborative_method = CollaborativeTMLEMethod(
    models=models,
    cross_fitting=folds,
    runtime=runtime,
    strategy="greedy",
    selection_folds=3,
    selection_inner_folds=2,
)
collaborative = effect.estimate(method=collaborative_method)
point = collaborative["ate"]
print(collaborative.summary(protocol="fingerprint"))
print()
plugin_low, plugin_high = point.plugin_interval
print(f"estimate:        {point.psi:.3f}")
print(f"inference:       {point.inference}")
print(f"plug-in standard error, a diagnostic:  {point.plugin_std_error:.4f}")
print(f"plug-in interval, a diagnostic:        ({plugin_low:.3f}, {plugin_high:.3f})")
print(f"population ATE:  {truth['ate']:.3f}")

Targeted maximum likelihood estimation
n = 2000; covariates = 3; P(A=1) = 0.512
causal estimand: average treatment effect, E[Y^a] - E[Y^reference]
identification: explicit-adjustment; E_W[E(transition_score | transition_navigation=a, W)] - E_W[E(transition_score | transition_navigation=0, W)] for a in [1]
required nuisances: outcome_regression, treatment_mechanism
identification assumptions: consistency: Y = Y^a when A = a; no interference: one unit's potential outcome does not depend on other units' treatment assignments; no unmeasured confounding: Y^a is independent of A given W; positivity: P(transition_navigation = a | W) > 0 almost surely for every supported treatment level a in [0, 1]
causal study protocol: schema 1; b81fd2923b8e38c8
outcome family: gaussian; fluctuation: logistic (iterative)
TMLE (in-sample nuisances): nuisances fitted in-sample (cross_fit=False)
propensity truncated to [0.01471, 0.9853]; ATT/ATC to [0.025, 0.975]
outcome scaled from [-6.926, 10.28] to [0, 1]
C-

**What this output tells you.** The summary repeats the estimand and the identification, and it
prints the protocol fingerprint. It names the construction as `TMLE (in-sample nuisances)`, which is
the ordinary point-treatment construction. The line `C-TMLE greedy selected candidate 2 of 4 for
ate` names the selector. Step 7 reads the selection from the nuisance report.

The estimate is 0.955, and the true ATE is 1.000. The estimate table has no interval column and no
p-value column. It has a `working-mechanism se` column instead, and the paragraph under it says
why. C-TMLE in `cleverly` reports a point estimate and a selection path. Its two plug-in values are
diagnostics.

| what the fit reports | value | what it means |
| --- | --- | --- |
| `psi` | 0.955 | the point estimate. The selection path in Step 7 stands with it |
| `inference` | `working_mechanism_plugin` | the fit declares that its reported curve is a diagnostic |
| `plugin_std_error` | 0.0440 | the plug-in standard error of the reported curve, under a name that claims no coverage |
| `plugin_interval` | (0.869, 1.041) | the Wald interval of that same curve, as a diagnostic |

The reported curve is the ordinary efficient influence curve at the candidate the search stopped
at. No result shows that this curve is the estimator's influence curve when the selected working
mechanism is not consistent for the treatment law. The trust section gives the sources.

Report the interval of the design-based TMLE fit in Step 8. The C-TMLE estimate is a check on it.
Read the plug-in pair as a spread of the curve the fit computed. Step 11 compares it with a robust
regression standard error.

## Step 7: read the selection path

The nuisance report retains the selection path. It lists the candidate assignment models in order
and marks the one that the cross-validated loss chose.


In [7]:
selection = collaborative.diagnostics.nuisance_models().selection
print(selection.summary())
print()
print(f"candidate path: {selection.path}")
print(f"selected covariates: {selection.selected_covariates}")
print(f"cv risk gap, k=1 minus k=0: {selection.cv_risk[1] - selection.cv_risk[0]:.5f}")

Collaborative TMLE selection
strategy = greedy; preorder = n/a; target = ate (ate); criterion = cross-validated penalized squared-error loss

k  covariates in g                                         steps  risk     cv risk     
-  ------------------------------------------------------  -----  -------  -------  ---
0  (intercept)                                             1      6.56613  6.63109     
1  social_support                                          1      6.56613  6.63098  <--
2  social_support, baseline_readiness                      2      6.56746  6.63428     
3  social_support, baseline_readiness, queue_lottery_draw  3      6.57264  6.65024     

selected g: social_support
left out: baseline_readiness, queue_lottery_draw. The selected candidate minimized targeted cross-validated penalized squared-error loss; this criterion does not determine why a covariate was left out

candidate path: ((), ('social_support',), ('social_support', 'baseline_readiness'), ('social_support

**What this output tells you.** The table lists four candidates for g. The greedy path starts
with the intercept only. It then adds `social_support`, `baseline_readiness`, and
`queue_lottery_draw`, in that order. The marker `<--` is on row 1, so the selected g holds
`social_support` alone.

The `cv risk` column decides the choice. The chosen candidate has the smallest value, 6.63098. The
candidate with all three variables has the largest value, 6.65024. The footer names the loss that
chose the candidate, and it says that this criterion does not determine why a covariate was left
out.

The first two candidates are nearly tied. The last line gives their `cv risk` gap, -0.00011. A
correct outcome regression leaves g almost nothing to improve, so the loss barely changes along the
path. A choice between near-tied candidates is fragile. The [technical
entry](../technical-reference/collaborative-tmle.md#validation-issues-special-to-this-method) lists
near-ties as an unresolved case for inference.

The selection also varies between draws. A review probe repeats this fit on 600 draws of this law
(`reviews/notebook-review/probes/collaborative-tmle-final/`). The selected g was empty on more than
half of them, and it held `social_support` alone on about a quarter. It contained the confounder on
about one in seven, and the draw on 9 of 600. The selector therefore left the draw out on 591 of
600 draws, as the design does.

On this draw, the selected variable is the one that moves the score and not assignment. The search
kept it, and it left out both the confounder and the instrument. The criterion is a loss on the
targeted outcome regression, and the report says so rather than assigning a role to any omission.

The other two columns describe the search. The `risk` column is the same penalized loss, computed
in sample. The `steps` column counts the targeting steps the search took before it reached that
candidate. When no added variable lowers the loss, the greedy search takes one more targeting step.
It then adds the best remaining variable even if the loss does not fall, so `risk` can rise along
the path. Only `cv risk` chooses the candidate.

The linear outcome regression is correctly specified for this law. A g that carries almost no
adjustment is then a legitimate choice. Step 9 explains why that choice alone does not test the
selector.

## Step 8: the reported fit, and an instrument in the assignment model

The code fits the plain TMLE on both designs of Step 5, with the same learners and the same
in-sample nuisances. The design-based fit is the reported analysis. The all-three fit puts the
queue draw into g, which is what a choice by predictive accuracy does. The code compares the fitted
propensity tails of the three fits. The support report describes how far the fitted propensities
reach toward 0 and 1. The code then prints the AUC of the all-three g and each ATE estimate.

An instrument in g pushes propensity scores toward 0 and 1 without removing confounding, so
precision falls. The clever covariate divides by the fitted propensity, so an extreme propensity
makes it large. [Point-treatment TMLE](../technical-reference/point-treatment-tmle.md) defines it.

The precision argument assumes exchangeability. If an unmeasured common cause remains, a strong
instrument can also amplify residual bias. C-TMLE does not turn the queue draw into an
instrumental-variable estimator.

In [8]:
plain_method = TMLEMethod(models=models, cross_fitting=folds, runtime=runtime)
reported = reported_effect.estimate(method=plain_method)
plain = effect.estimate(method=plain_method)


def tails(result):
    support = result.diagnostics.support()
    return {
        "share of g below 0.1": support.tail_mass[0.1]["below"],
        "share of g above 0.9": support.tail_mass[0.1]["above"],
        "truncated fraction": support.truncated["fraction"],
        "treated ESS / n": support.effective_sample_size["treated"]["ratio"],
        "control ESS / n": support.effective_sample_size["control"]["ratio"],
        "max |clever covariate|": support.clever_covariate_max["mean"],
    }


tail_table = pd.DataFrame(
    {
        "design-based TMLE": tails(reported),
        "all-three TMLE": tails(plain),
        "collaborative TMLE": tails(collaborative),
    }
)
print(tail_table.round(4))
plain_auc = plain.diagnostics.nuisance_models()["propensity"].metrics["auc"]
print(f"all-three TMLE propensity AUC: {plain_auc:.3f}")
print()


def show(label, result):
    point = result["ate"]
    if point.inference == "influence_curve":
        low, high = point.ci
        spread = f"se={point.std_error:6.4f}  CI=({low:.3f}, {high:.3f})"
    else:
        low, high = point.plugin_interval
        spread = f"plug-in se={point.plugin_std_error:6.4f}  plug-in ({low:.3f}, {high:.3f})"
    print(f"{label:22s} psi={point.psi:6.3f}  {spread}")


show("design-based TMLE", reported)
show("all-three TMLE", plain)
show("collaborative TMLE", collaborative)
print("the plug-in rows are diagnostics, not inference")
print(f"population ATE: {truth['ate']:.3f}")

                        design-based TMLE  all-three TMLE  collaborative TMLE
share of g below 0.1               0.0000          0.1040              0.0000
share of g above 0.9               0.0000          0.1140              0.0000
truncated fraction                 0.0000          0.0160              0.0000
treated ESS / n                    0.9030          0.4719              0.9985
control ESS / n                    0.8915          0.4499              0.9984
max |clever covariate|             6.3471         43.7542              2.3819
all-three TMLE propensity AUC: 0.844

design-based TMLE      psi= 1.012  se=0.0466  CI=(0.921, 1.104)
all-three TMLE         psi= 0.881  se=0.0626  CI=(0.758, 1.004)
collaborative TMLE     psi= 0.955  plug-in se=0.0440  plug-in (0.869, 1.041)
the plug-in rows are diagnostics, not inference
population ATE: 1.000


**What this output tells you.** The design-based fit is the reported analysis. It estimates the
ATE at 1.012, with a standard error of 0.0466 and the interval (0.921, 1.104). The interval
contains the true value of 1.000 on this draw. Its g has no rows in either tail. Its
effective-sample ratios are 0.9030 in the treated arm and 0.8915 in the control arm, because
readiness still moves this g.

The all-three fit shows the failure mode. A share of 0.1040 of its fitted propensities lies below
0.1, and a share of 0.1140 lies above 0.9. Its largest clever covariate is 43.7542. Its
effective-sample ratios are 0.4719 and 0.4499. In this law the draw moves assignment more strongly
than readiness does.

This g has an AUC of 0.844, so it predicts the offer well. The draw adds to that accuracy and
removes no confounding, because the draw has no path to the score in this law.

The all-three fit reports a standard error of 0.0626 and the interval (0.758, 1.004). The
collaborative fit has no rows in either tail, and its effective-sample ratios are 0.9985 and
0.9984. On this draw, its g holds one variable that does not move assignment, so every row in an
arm carries nearly the same weight. Its estimate of 0.955 lies inside the reported interval. Its
plug-in diagnostic is 0.0440, with the plug-in spread (0.869, 1.041).

The review probe of Step 7 repeats the three fits on 600 draws. Its numbers describe this
configuration only.

| over 600 draws | design-based TMLE | all-three TMLE | C-TMLE |
| --- | --- | --- | --- |
| empirical SD of the estimate | 0.048 | 0.069 | 0.055 |
| mean standard error or plug-in value, over that SD | 0.97 | 0.92 | 0.81 |
| draws on which the interval or the plug-in spread contains 1.000 | 572 | 555 | 532 |

The reported interval covered near the nominal 95%, and its standard error matched the spread. The
all-three interval covered less often. Its g holds the draw, so its propensities reach toward 0
and 1. Its standard error then understates the spread at this size.

C-TMLE agrees with the design. It left the draw out of g on 591 of 600 draws, and its estimate lay
inside the reported interval on 598. Its spread is about four fifths of the all-three spread, but
it is larger than the spread of the reported fit. The selector is a safeguard when the design
gives no knowledge of an instrument. It is not a precision gain over that knowledge.

Do not refit a plain TMLE on the set that C-TMLE selected, here `social_support`, and report its
interval. That is post-selection inference. The selected set changes from draw to draw, and no
result gives coverage to an interval that conditions on it. The design-based set follows from the
protocol, which the program fixed before the data. It is not a selection.

The plug-in value understates even the C-TMLE spread on this law. The selected g barely moves, so
the plug-in curve ignores the adjustment that the outcome regression performs. The plug-in value
was below the HC0 standard error of Step 11 on 595 of 600 draws.

Step 9 explains why the comparison of these fits does not test the selector.

## Step 9: a control that discriminates

On this draw, with a correctly specified outcome model, Step 7 showed the selector stop at a
variable that does not move assignment. That choice can minimize the targeted cross-validated loss. It does not show
that the search can tell a confounder from an instrument. A selector that always chose an
adjustment-free g would give the same result.

Use a deliberate stress control where selecting nothing is wrong. The code reduces the outcome model
to a constant with `DummyRegressor`, so the assignment model must carry the adjustment. This tests
the selector. It is not a recommended production outcome model.

`replace` keeps every setting of the Step 6 and Step 8 methods and changes only the learners. The
code prints both estimates, the propensity tails of both fits, and the selection path.

In [9]:
weak_models = ModelSpec(
    outcome_learner=DummyRegressor(),
    treatment_learner=LogisticRegression(max_iter=1000, random_state=44),
)
weak_plain = effect.estimate(method=replace(plain_method, models=weak_models))
weak_collaborative = effect.estimate(method=replace(collaborative_method, models=weak_models))
weak_selection = weak_collaborative.diagnostics.nuisance_models().selection
ratio = weak_collaborative["ate"].plugin_std_error / weak_plain["ate"].std_error
show("constant Q, plain", weak_plain)
show("constant Q, C-TMLE", weak_collaborative)
print(f"plug-in over plain standard error: {ratio:.3f}")
print(f"population ATE: {truth['ate']:.3f}")
print()
weak_tail_table = pd.DataFrame(
    {"constant Q, plain": tails(weak_plain), "constant Q, C-TMLE": tails(weak_collaborative)}
)
print(weak_tail_table.round(4))
print()
print(weak_selection.summary())

constant Q, plain      psi= 1.077  se=0.2343  CI=(0.618, 1.537)
constant Q, C-TMLE     psi= 1.017  plug-in se=0.0923  plug-in (0.836, 1.197)
plug-in over plain standard error: 0.394
population ATE: 1.000

                        constant Q, plain  constant Q, C-TMLE
share of g below 0.1               0.1040              0.0000
share of g above 0.9               0.1140              0.0000
truncated fraction                 0.0160              0.0000
treated ESS / n                    0.4719              0.9030
control ESS / n                    0.4499              0.8915
max |clever covariate|            43.7542              6.3471

Collaborative TMLE selection
strategy = greedy; preorder = n/a; target = ate (ate); criterion = cross-validated penalized squared-error loss

k  covariates in g                                         steps  risk     cv risk     
-  ------------------------------------------------------  -----  -------  -------  ---
0  (intercept)                            

**What this output tells you.** With a constant Q, the selected g is
`baseline_readiness, social_support`. The footer lists `queue_lottery_draw` as left out. On this
draw the selector keeps the confounder and leaves the instrument out.

The path table shows the change in the loss. Adding `baseline_readiness` lowers the `cv risk` from
26.2934 to 24.9909. Adding `queue_lottery_draw` to the other two raises it from 24.8761 to 24.9728.
The search also keeps `social_support`, which lowers the `cv risk` from 24.9909 to 24.8761. It does
not move assignment in this law, so it adds little to g.

The tail table shows what the constant-Q search does to g. The plain g holds all three variables,
so it has the tails of the all-three fit in Step 8. The C-TMLE g has no rows in either tail. Its effective-sample
ratios are 0.9030 and 0.8915, and its largest clever covariate is 6.3471. Readiness still moves
this g, so the ratios are below 1.

The C-TMLE plug-in standard error is 0.0923, and the plain standard error is 0.2343. Their ratio is
0.394. It compares a diagnostic with a standard error, so it does not measure the precision gain.
The plain interval and the C-TMLE plug-in spread both contain 1.000 on this draw.

The plain constant-Q fit is a stress comparator, not a reference. In the review probe, its estimate
ran about 0.1 above the truth on average over 600 draws, and its interval covered on 513.

The selected constant-Q fit had about a third of the plain spread. In this configuration its plug-in
value overstated that spread. In the cross-fitted bounded twin of the registered study, the plug-in
value understated it. The direction of the plug-in error depends on the configuration. The
constant-Q C-TMLE estimates are heavy-tailed, so read these repeated-draw values as rough.

Over the same 600 draws, the constant-Q search kept the confounder on 599 and left the draw out on
565. The selection result does not prove that either variable has its declared causal role.

The [registered
study](../technical-reference/method-evidence/selector-based-point-treatment-c-tmle.md) runs a
similar control on a bounded twin of this law. It cross-fits at n = 1,500 over 800 draws, and it
scores bias and error, not the selected set. There, the bias of a selector forced to stop at the
empty path lies entirely outside the equivalence margin, which is what that control requires. The
free selector's root-mean-square error stays well below the forced one's, so the search does work on
that law. The free selector's own bias interval reaches past its margin, and the study publishes
that cell red.

## Step 10: diagnostics, what the fit can check

Start with the combined assessment of the Step 6 fit. The code prints each assessment row that ran
or that runs on request, with its status. It then prints the fitted range of g and the truncated
fraction from the support report. It also prints the nuisance report, with the role that report
gives to g.

In [10]:
assessment = collaborative.assess()
support = assessment.report("support")
nuisance = assessment.report("nuisance_models")
ledger = assessment.to_frame()
ran = ledger[~ledger["status"].isin(["unavailable", "not_applicable"])]
print(ran[["surface", "check", "status"]].to_string(index=False))
print(f"needs attention: {tuple(item.name for item in assessment.attention)}")
print()
quantiles = support.propensity_quantiles["overall"]
print(f"fitted g range: {quantiles[0.0]:.4f} to {quantiles[1.0]:.4f}")
print(f"truncated fraction: {support.truncated['fraction']:.4f}")
print()
print(f"treatment role: {nuisance.treatment_role}")
print(nuisance.summary())
print()
print(f"propensity AUC: {nuisance['propensity'].metrics['auc']:.3f}")
print(f"selected covariates: {nuisance.selection.selected_covariates}")

    surface                 check    status
 validation       score_equations    passed
 validation               support completed
 validation       nuisance_models completed
diagnostics      truncation_curve  deferred
diagnostics                refute  deferred
sensitivity simulated_confounding  deferred
needs attention: ()

fitted g range: 0.4434 to 0.5814
truncated fraction: 0.0000

treatment role: collaborative_working_model
Nuisance model diagnostics
----------------------------------------
model       auc     brier   log_loss  r2      mse     cal_slope  cal_se  reg_slope
----------  ------  ------  --------  ------  ------  ---------  ------  ---------
propensity  0.5158  0.2495  0.6921    -       -       1.0022     0.5704  -        
outcome     -       -       -         0.8002  0.0033  -          -       1.0000   
Calibration slopes are in sample and carry no test. An unpenalized logistic model with an intercept has an in-sample slope of 1 by construction.

The propensity metri

**What this output tells you.** Read the four parts in order.

| output part | what it shows on this draw |
| --- | --- |
| the status table and `needs attention` | the score-equation check passed, and nothing needs attention. `support` and `nuisance_models` are `completed`: the calculation ran. `completed` is not a pass. The `deferred` rows run on request |
| `fitted g range` and `truncated fraction` | every fitted propensity lies between 0.4434 and 0.5814, and no row is truncated |
| nuisance model diagnostics | the treatment role is `collaborative_working_model`, and the propensity AUC is 0.516 |

Step 11 runs `simulated_confounding`, a `deferred` row here, on this fit. The omitted-variable
bound needs the representer of an assignment model that the fit estimates. Step 11 therefore reads
that bound on the reported design-based fit. The nuisance report and its verdict both state that
this fit reports no confidence interval and no p-value.

The selected model carries one variable that does not move assignment, so its AUC sits near 0.5 and
its calibration slope near 1. This fit is in sample, so the report tests no slope, and the line
under its table says so. Neither number describes assignment given the complete adjustment
set. They describe the working denominator only.
[Nuisance model quality](../technical-reference/validation-methods.md#nuisance-model-quality) lists
the two claims that this role drops.

The support report describes the selected denominator only. Compare it with the all-three fit in Step 8
to see the tails that selection removed. Neither report assigns a causal role to an omitted
variable.

## Step 11: sensitivity, what the fit cannot check

Sensitivity analysis asks how strong an unmeasured confounder would need to be to change the
conclusion. The [results guide](../user-guide/results-assessment.md#sensitivity-analysis) introduces
it. Sensitivity analysis cannot tell whether the selector chose a useful assignment model.

The code runs one analysis for each fit.

| fit | analysis | what it reads |
| --- | --- | --- |
| design-based TMLE, the reported fit | the omitted-variable elements and the robustness value | the representer of the design-based assignment model, which this fit estimates |
| all-three TMLE | the same rows, for comparison | the representer of the assignment model on all three variables |
| C-TMLE | `simulated_confounding`, a simulated common-cause stress surface | how the estimate moves when one latent variable flips treatment or shifts the score |

The
[stress-surface section](../technical-reference/validation-methods.md#simulated-common-cause-stress-surface)
describes the surface. The grid has two settings on each axis. A treatment strength of 0.1 flips
the treatment of each row whose latent value is in the top 10% of the standard normal. An outcome
strength of 0.5 subtracts 0.5 times the latent value from the score. The score has no declared
range, so the outcome axis can run on it.

The code also fits a linear regression of the score on the offer and all three covariates. That
regression gives the offer coefficient and its robust (HC0) standard error. HC0 stays valid when
the residual variance differs across rows. The code prints that standard error beside the plug-in
diagnostic of the collaborative fit.

In [11]:
from cleverly.learners import thread_limit
from cleverly.sensitivity import ConfounderStrengthGrid


def bound_rows(result):
    fit_assessment = result.assess()
    elements = fit_assessment.report("elements")
    robustness = fit_assessment.report("robustness_value")
    return {
        "estimate": result["ate"].psi,
        "sigma2": elements.sigma2,
        "nu2": elements.nu2,
        "robustness value": robustness["rv"],
        "confidence-limit value": robustness["rva"],
    }


bound_table = pd.DataFrame(
    {"design-based TMLE": bound_rows(reported), "all-three TMLE": bound_rows(plain)}
)
print(bound_table.round(3))
print()

surface = collaborative.sensitivity.simulated_confounding(
    estimand="ate",
    grid=ConfounderStrengthGrid(treatment=(0.0, 0.1), outcome=(0.0, 0.5)),
)
print(surface.summary())
print()

offer, score = frame["transition_navigation"], frame["transition_score"]
with thread_limit():
    full = frame[["transition_navigation", *covariates]]
    residual = score - LinearRegression(n_jobs=1).fit(full, score).predict(full)
    offer_part = offer - LinearRegression(n_jobs=1).fit(frame[covariates], offer).predict(
        frame[covariates]
    )
coefficient = (offer_part * score).sum() / (offer_part**2).sum()
robust_se = (offer_part**2 * residual**2).sum() ** 0.5 / (offer_part**2).sum()
print(
    f"C-TMLE estimate and plug-in standard error: {collaborative['ate'].psi:.3f}, "
    f"{collaborative['ate'].plugin_std_error:.4f}"
)
print(f"regression offer coefficient and HC0 error: {coefficient:.3f}, {robust_se:.4f}")

                        design-based TMLE  all-three TMLE
estimate                            1.012           0.881
sigma2                              0.971           0.968
nu2                                 4.486          11.963
robustness value                    0.381           0.227
confidence-limit value              0.357           0.204



Simulated common-cause stress surface for 'ate'
original estimate: 0.95497
seeds: root=44, latent=4136545821, refit=44
cross-fitting: 1 draw, no repeat aggregation
target measure: unweighted
target population: baseline
conditioning arm: None
baseline stratum: None
strata columns: ()
association population: full_fitted_population
calibration population: full_fitted_population
refit population: full_fitted_population
Treatment is flipped in the declared upper latent-normal tail.
Gaussian outcome subtracts signed strength times the shared latent value.

treatment strength  outcome strength  estimate  movement (estimate difference)  induced association  population fraction
------------------  ----------------  --------  ------------------------------  -------------------  -------------------
0                   0                 0.95497   +0                              -0.0480              1.0000             
0                   0.5               1.0232    +0.068278                       

**What this output tells you.** The reported fit has a robustness value of 0.381 and a
confidence-limit value of 0.357. Its residual variance `sigma2` is 0.971, and its `nu2` is 4.486.
The all-three fit has a robustness value of 0.227, a confidence-limit value of 0.204, and a `nu2`
of 11.963. The [omitted-variable
bounds](../technical-reference/validation-methods.md#omitted-variable-bounds-robustness-value-benchmark-and-contours)
define `nu2` from the Riesz representer given the adjustment set.

`nu2` explains the difference. The representer of the ATE divides by the fitted propensity, so the
extreme propensities of the all-three g inflate its second moment. A larger `nu2` widens the bias
bound for each confounder strength, so a weaker confounder can reach the null. The draw in g
therefore lowers the robustness value as well as the precision. In the review probe, the
robustness value was larger on the design-based fit on 596 of 600 draws.

The bound needs a consistent assignment model for its adjustment set. The true g of the
design-based set averages a logistic curve over the draw, so a logistic model approximates it. The
probe script `truth.py` integrates the true g, and its population `nu2` is 4.357. Over the 600
probe draws, the fitted `nu2` stayed between 4.164 and 4.702.

The surface starts from the C-TMLE estimate, 0.95497.

| treatment strength | outcome strength | estimate | movement | induced association |
| --- | --- | --- | --- | --- |
| 0 | 0 | 0.95497 | 0 | -0.0480 |
| 0 | 0.5 | 1.0232 | +0.068278 | -0.0480 |
| 0.1 | 0 | 0.63882 | -0.31615 | -0.0252 |
| 0.1 | 0.5 | 0.60972 | -0.34525 | -0.0252 |

The induced association is the correlation between the latent value and the treatment. The latent
value is drawn independently of the data. At treatment strength 0, its association of -0.0480 is
therefore a chance correlation on this draw. The outcome strength alone moves the estimate by
+0.068278 through that small correlation.

The treatment flips move the estimate down by about a third. A flip turns an offer into usual
support, or the reverse, whatever the patient's score. That pulls the contrast toward zero. The
summary reads a near-zero association the same way: as misclassification, not as confounding.

Read this surface as a stress test of the collaborative fit under one latent cause. It is
qualitative. It is not a bound, and it is not an adjusted interval. Choose strengths that the
causal review of the program can defend before you read a movement.

The last two lines compare the plug-in diagnostic with a robust regression standard error. The
plug-in diagnostic of the collaborative fit is 0.0440. The regression gives the same estimate,
0.955, with an HC0 standard error of 0.0545.

The plug-in value is smaller. In this law, that holds on almost every draw: the review probe found
it smaller on 595 of 600 draws. Over those draws, the HC0 standard error averaged 0.98 of the spread
of the C-TMLE estimate. Neither number is a confidence statement for this estimator.

## How far to trust this

Three limits belong in every report of a collaborative fit on this page.

**Report the design-based TMLE interval for the ATE.** The design-based set follows from the
protocol, which records the hash that fixes the queue draw. In the review probe, the reported
interval covered on 572 of 600 draws. C-TMLE in `cleverly` reports a point estimate, a selection
path, and two plug-in diagnostics. No result shows that the reported curve is this estimator's
influence curve when the selected mechanism is not consistent for the treatment law. Use the
C-TMLE estimate as a check of the design, and do not report an interval refitted on its selected
set.

**The plug-in diagnostic understates the spread on this law.** The near-constant working model of
Step 7 ignores the adjustment that the outcome regression performs. Step 11 shows the plug-in value
of 0.0440 beside the HC0 standard error of 0.0545 for the same coefficient. In the review probe, the
plug-in value averaged 0.81 of the empirical spread. The [technical
entry](../technical-reference/collaborative-tmle.md#validation-issues-special-to-this-method)
records the limit and the sources.

**The nuisances here are fitted in sample.** Every fit on this page reuses each row for both
fitting and prediction. The reported interval needs the data-reuse condition that
[CV-TMLE](../technical-reference/cv-tmle.md#what-this-solves) states. The linear learners of this
page satisfy it. The synthetic score also has no documented range to declare as `q_bounds`. With
flexible learners and a real questionnaire, declare its published range and cross-fit.

| layer | establishes | does not establish |
| --- | --- | --- |
| the combined assessment | which cached checks need attention and which costly operations did not run | selection uncertainty or the causal role of a candidate variable |
| the support report | how far the selected denominator reaches into the tails | that the selected model is the right one |
| the nuisance report | selected-model metrics, model role, and the retained selection | whether low AUC means limited confounding after collaborative selection |
| the retained selection path | which candidates the search considered and selected on this draw | that the search selects the same set on another draw, or a set on which to report an interval |
| the plug-in standard error and interval | the spread of the curve this fit computed at the selected candidate | coverage for this estimator. On this law the plug-in value understates the spread |
| the design-based TMLE interval | inference for the ATE on the design-based set. It covered on 572 of 600 probe draws | that the queue draw meets the exclusion restriction. The protocol records its design basis, and the data cannot check it |
| the all-three TMLE fit | what the draw in g does to the tails, the spread, and the robustness value | an interval to report |
| the sensitivity analysis | on the reported fit, a robustness value for the design-based set when its assignment model is consistent. On the collaborative fit, how the estimate moves under one simulated latent cause | a bound on the collaborative fit, or a bound that accounts for an inconsistent assignment model |

No registered study covers the configuration of this page: the Gaussian score, the linear outcome
regression, and the in-sample fits. The review probe in
`reviews/notebook-review/probes/collaborative-tmle-final/` covers it, and the probe is not a
registered study. The table lists the registered cells that cover parts of the construction.

| registered study and cited cell | law | nuisance construction | outcome type | fold layer | role | what the cell shows |
| --- | --- | --- | --- | --- | --- | --- |
| [selector-based C-TMLE](../technical-reference/method-evidence/selector-based-point-treatment-c-tmle.md), `selector_necessity/collaborative` | another law: the bounded twin of this page's law, at n = 1,500 | Q a fitted constant; g a fitted logistic model on the set the greedy search selects | continuous, bounded in 0 to 1 | cross-fitted, five folds | positive | a root-mean-square error of 0.241 times the forced selector's. The 99% bias interval reaches 0.0037, past the margin of 0.0030, so the study publishes the cell red. The plug-in value is 0.80 of the empirical spread |
| the same study, `selector_necessity/empty_control` | the same | the same, with g forced to the empty set | the same | the same | control | a bias interval of 0.0495 to 0.0511, outside its margin of 0.0022, as the control requires |
| [outcome-adaptive C-TMLE](../technical-reference/method-evidence/outcome-adaptive-point-treatment-c-tmle.md), `crossfit_overfitting/cross_fitted_oat` | another law: the bounded law of `navigation_data`, at n = 500 | Q a fully grown regression tree; g a fitted logistic model of the offer on the fitted Q, the shared outcome-adaptive design | continuous, bounded in 0 to 1 | cross-fitted, five folds | positive | coverage of 0.92, with a standard error of 0.99 times the empirical spread |
| the same study, `crossfit_overfitting/in_sample_control` | the same | the same | the same | in sample | control | coverage of 0.54, with a standard error of 0.47 times the empirical spread. An in-sample flexible Q breaks the interval |

The selector-based study also validates the greedy selector against R `ctmle` with logistic GLMs.
Its agreement rows use a binary-outcome law.

## Where to go next

Collaborative TMLE addresses *selection*. If your worry is the *inference* instead, because you
expect one nuisance to be inconsistent however you choose it, read [DR-TMLE](dr-tmle.ipynb). If the
adjustment set is small and you would include all of it, the plain
[point-treatment TMLE](point-treatment-tmle.ipynb) is the right entry.

C-TMLE in `cleverly` covers point-treatment targets. The
[technical entry](../technical-reference/collaborative-tmle.md#variations) lists them and the
variations. The
[examples index](index.md#the-program) lists every tutorial in the program.